// smartOpts/smartOpts.js
let smartOpts = function () {
  this.optIn = function (option, list = this.optionsMutable) {
    if (list.indexOf(option) >= 0) {
      return true
    } else {
      return false
    }
  },
  this.optsIn = function (options, list = this.optionsMutable) {
    for (let option of options) {
      if (list.indexOf(option) >= 0) {
        // return true
      } else {
        return false
      }
    }
    return true
  },
  this.anyOptsIn = function (options, list = this.optionsMutable) {
    for (let option of options) {
      if (list.indexOf(option) >= 0) {
        return true
      }
    }
    return false
  },
  this.optIndex = function (option, list = this.optionsMutable) {
    return list.indexOf(option)
  },
  this.pushOpt = function (option, list = this.optionsMutable) {
    if (!this.optIn(option)) {
      list.push(option)
    }
  },
  this.pushOpts = function (options, list = this.optionsMutable) {
    for (let option of options) {
      this.pushOpt(option)
    }
  },
  this.popOpt = function (option, list = this.optionsMutable) {
    if (this.optIn(option)) {
      list.splice(this.optIndex(option), 1)
    }
  },
  this.popOpts = function (options, list = this.optionsMutable) {
    for (let option of options) {
      this.popOpt(option)
    }
  },
  this.toggleOpt = function (option, list = this.optionsMutable) {
    if (this.optIn(option)) {
      this.popOpt(option)
    } else {
      this.pushOpt(option)
    }
  },
  this.toggleOpts = function (options, list = this.optionsMutable) {
    for (let option in options) {
      if (this.optIn(option)) {
        this.popOpt(option)
      } else {
        this.pushOpt(option)
      }
    }
  }
}

module.exports = new smartOpts()
