// secrets/secrets.js
require('dotenv').config()

let secrets = {
	mongodb: {
		auth: true,
		db: process.env.MONGODB_DB,
		server: `${process.env.MONGODB_CLUSTER}.nhb33.mongodb.net`,
		ip: `${process.env.MONGODB_CLUSTER}.nhb33.mongodb.net`,
		username: process.env.MONGODB_USER,
		password: process.env.MONGODB_PWD
	},
	sessionSecret: process.env.SESSION_SECRET,
	facebookApp: {
		id: process.env.FACEBOOK_ID,
		secret: process.env.FACEBOOK_SECRET
	},
	firebase: {
	},
	google: {
		CLIENT_ID: process.env.GOOGLE_CLIENT_ID,
	},
	email:{
		email: process.env.EMAIL,
		password: process.env.EMAIL_PASSWORD
	},

}

secrets.monk = secrets.node214 = secrets.mongodb

secrets.google.firebaseUrl = process.env.FIREBASE_URL
secrets.firebase = {
	"type": "service_account",
	"project_id": process.env.FIREBASE_PROJECT_ID,
	"private_key_id": process.env.FIREBASE_PK_ID,
	"private_key": process.env.FIREBASE_PK.replace(/\\n/g, '\n'),
	"client_email": process.env.FIREBASE_CLIENT_EMAIL,
	"client_id": process.env.FIREBASE_CLIENT_ID,
	"auth_uri": "https://accounts.google.com/o/oauth2/auth",
	"token_uri": "https://oauth2.googleapis.com/token",
	"auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
	"client_x509_cert_url": process.env.FIREBASE_CLIENT_X509
}


secrets.mongodb.uri = "mongodb+srv://" + (secrets.mongodb.auth ? secrets.mongodb.username + ":" + secrets.mongodb.password + "@" : '') + secrets.mongodb.server + "/" + secrets.mongodb.db + "?retryWrites=true&w=majority"

secrets.mdb = secrets.mongodb

module.exports = secrets
